### PR TITLE
Update: check renaming identifier in object destructuring (fixes 12827)

### LINKED ISSUE #12827: `id-length` should warn about the renamed identifier length in the destructuring assignment.
**Tell us about your environment**

* **ESLint Version:** v6.8.0
* **Node Version:** v12.13.0
* **npm Version:** v6.12.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

[Online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgaWQtbGVuZ3RoOiBcImVycm9yXCIqL1xuXG5jb25zdCB7IHg6IHMgfSA9IGJhcjsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJydWxlcyI6e30sImVudiI6eyJicm93c2VyIjpmYWxzZX19fQ==)
```js
/*eslint id-length: "error"*/

const { x: s } = bar;
```

**What did you expect to happen?**

It should warn about the identifier`s` length rather than renamed `x`.

**What actually happened? Please include the actual, raw output from ESLint.**

It warns about the length of 'x';
```
3:9- Identified name 'x' is too short(<2).(id-length)
```
**Are you willing to submit a pull request to fix this bug?**

Yes.

### PR DESCRIPTION
**Pull request by yeonjuan**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).
- [x] The team has reached consensus on the changes proposed in this pull request. If not, I understand that the evaluation process will begin with this pull request and won't be merged until the team has reached consensus.

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

```js
const { renamed: renaming} = foo;
```
Changed this rule to warn the length of the renaming identifier rather than a renamed identifier.

#### Is there anything you'd like reviewers to focus on?

Fixes #12827

### CODE REVIEW

File: lib/rules/id-length.js [hunk 1]
```diff
@@ -83,6 +84,13 @@ module.exports = {
...
             },
             Property: properties && function(parent, node) {
+
+                if (parent.parent.type === "ObjectPattern") {
+                    return (
+                        parent.value !== parent.key && parent.value === node ||
+                        parent.value === parent.key && parent.key === node
+                    );
+                }
                 return parent.key === node;
```
Comment (mdjermanovic) [inline lib/rules/id-length.js]: This retains the actual behavior, and that is to entirely ignore all object expression and object pattern properties when `"properties"` option is set to `"never"`, but we should probably revisit that at some point. 

The intention of this option is, most likely, to allow creating properties with invalid names. But, it also allows this:

```js
/*eslint id-length: ["error", { "min": 2, "properties": "never" }]*/

const { a } = obj; // no error, this is likely a false negative

const { prop: b } = obj; // no error, this is even more likely a false negative
```

Additionally, and unrelated to the previous, this is probably a bug:

```js
/*eslint id-length: ["error", { "min": 2, "properties": "always" }]*/

const foo = { [a]: 1 }; // error on a, this is likely a false positive

const bar = { [a + b]: 1 }; // no error
```

File: docs/rules/id-length.md [hunk 2]
```diff
@@ -31,6 +31,7 @@ var myObj = { a: 1 };
 class x { }
 class Foo { x() {} }
 function foo(...x) { }
+function foo({x}) { }
 var { x } = {};
 var { x: a} = {};
```
Comment (mdjermanovic) [inline docs/rules/id-length.md]: It might be good to avoid confusion about which one is invalid, maybe we could rename `x` to have the correct length?

File: tests/lib/rules/id-length.js [hunk 3]
```diff
@@ -126,27 +132,140 @@ ruleTester.run("id-length", rule, {
...
+                tooShortError
+            ]
+        },
+        {
+            code: "var { x: a} = {};",
+            parserOptions: { ecmaVersion: 6 },
+            errors: [
+                tooShortError
+            ]
+        },
```
Comment (mdjermanovic) [inline tests/lib/rules/id-length.js]: It might be nice to also add one with `{ a: a }`.

File: tests/lib/rules/id-length.js [hunk 4]
```diff
@@ -50,15 +51,20 @@ ruleTester.run("id-length", rule, {
...
         { code: "var { prop } = {};", parserOptions: { ecmaVersion: 6 } },
-        { code: "var { prop: a } = {};", parserOptions: { ecmaVersion: 6 } },
+        { code: "var { a: foo } = {};", options: [{ min: 3 }], parserOptions: { ecmaVersion: 6 } },
+        { code: "var { prop: foo } = {};", options: [{ max: 3 }], parserOptions: { ecmaVersion: 6 } },
+        { code: "var { longName: foo } = {};", options: [{ min: 3, max: 5 }], parserOptions: { ecmaVersion: 6 } },
+        { code: "var { foo: a } = {};", options: [{ exceptions: ["a"] }], parserOptions: { ecmaVersion: 6 } },
         { code: "var { prop: [x] } = {};", parserOptions: { ecmaVersion: 6 } },
+        { code: "var { a: { b: { c: longName } } } = {};", parserOptions: { ecmaVersion: 6 } },
+        { code: "({ a: obj.x.y.z } = {});", options: [{ properties: "never" }], parserOptions: { ecmaVersion: 6 } },
+        { code: "var { a: [x]} = {};", parserOptions: { ecmaVersion: 6 } },
```
[the remark this answers is not in the document]
Comment (yeonjuan) [reply tests/lib/rules/id-length.js]: > Also, I believe it will become invalid after #12839, so the build might fail when both PRs get merged.

@mdjermanovic 
Yes, right. it will conflict, and I'll resolve that Thanks 👍

File: tests/lib/rules/id-length.js [hunk 5]
```diff
@@ -126,27 +132,140 @@ ruleTester.run("id-length", rule, {
...
+                tooShortError
+            ]
+        },
+        {
+            code: "function foo({x: a}) { }",
+            parserOptions: { ecmaVersion: 6 },
+            errors: [
+                tooShortError
+            ]
+        },
```
[the remark this answers is not in the document]
Comment (yeonjuan) [reply tests/lib/rules/id-length.js]: > I think it would be good to add data (maybe also location?) to some of the new tests, at least to those with two or more potentially invalid names, to confirm that this reports a, not x.

I agree! it makes sense. :) thanks

File: lib/rules/id-length.js (hunk 1 above)
Comment (yeonjuan) [reply lib/rules/id-length.js]: @mdjermanovic 
Thanks 👍 for the review. I think both can be fixed. I'll try : )

File: tests/lib/rules/id-length.js [hunk 6]
```diff
@@ -176,6 +313,30 @@ ruleTester.run("id-length", rule, {
...
-        { code: "var x = 1;", options: [{ properties: "never" }], errors: [tooShortError] }
+        { code: "var x = 1;", options: [{ properties: "never" }], errors: [tooShortError] },
+        {
+            code: "var {x} = foo;",
+            options: [{ properties: "never" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [
+                tooShortError
+            ]
+        },
```
[the remark this answers is not in the document]
Comment (yeonjuan) [reply tests/lib/rules/id-length.js]: > properties: "never" in this rule seems to be a mix of properties: "never" and "ignoreDestructuring": true options from other naming convention rules, like camelcase and id-match.

Oh, I got it. I'll change. Thanks :)

**Comment by mdjermanovic** [comment]
Checked examples from the docs:

* `var { a: [x]} = {};` examples will be valid after this PR (without the other PR), maybe we should remove them from invalid sections, in this PR.
* `var { prop: a } = {};` in section `Examples of incorrect code for this rule with the { "min": 4 } option:` is now invalid.
* In the same section, `var foo = function (event) { /* do stuff */ };` is also invalid. that was probably a mistake in the docs.
* In the same section, `obj.element = document.body;` and `({ prop: obj.name } = {});` are both invalid on `obj`. That was probably broken in the code. Not sure what should we do with this, maybe leave for another PR.

File: tests/lib/rules/id-length.js [hunk 7]
```diff
@@ -126,27 +138,155 @@ ruleTester.run("id-length", rule, {
...
+        {
+            code: "function foo({ a: b }) {};",
+            options: [{ exceptions: ["a"] }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [
+                tooShortError
+            ]
+        },
+        {
+            code: "function foo({ a: { b: { c: d, e } } }) { }",
```
Comment (kaicataldo) [inline tests/lib/rules/id-length.js]: Not a blocker for this PR, but it seems like it would be useful to include the location information in the errors for tests like this, to ensure that it's warning about the correct identifier.

File: docs/rules/id-length.md [hunk 8]
```diff
@@ -89,8 +91,7 @@ class x { }
 class Foo { x() {} }
 function foo(...x) { }
 var { x } = {};
-var { x: a} = {};
-var { a: [x]} = {};
```
Comment (kaicataldo) [inline docs/rules/id-length.md]: Is this a regression in this PR? I see [that this is going to be worked on in a separate PR](https://github.com/eslint/eslint/pull/12881#issuecomment-584330660). Do we have an issue outlining our plan for this (just in case someone else needs to pick this work up)?

**Comment by kaicataldo** [comment]
Thanks for working on this!

File: tests/lib/rules/id-length.js (hunk 7 above)
Comment (yeonjuan) [reply tests/lib/rules/id-length.js]: @kaicataldo 
you are right! 👍I added. thanks

File: docs/rules/id-length.md (hunk 8 above)
Comment (yeonjuan) [reply docs/rules/id-length.md]: > Is this a regression in this PR?

Yes. The case will not be warned after this PR merged because the actual version warns about `a` rather than `x`. 

* [Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgaWQtbGVuZ3RoOiBbXCJlcnJvclwiLCB7IFwibWluXCI6IDQgfV0qL1xuLyplc2xpbnQtZW52IGVzNiovXG5cbnZhciB7IGE6IFt4XX0gPSB7fTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJydWxlcyI6e30sImVudiI6eyJicm93c2VyIjp0cnVlLCJlczIwMjAiOnRydWUsIm5vZGUiOnRydWUsImVzMjAxNyI6dHJ1ZSwiZXM2Ijp0cnVlfX19)

```js
/*eslint id-length: ["error", { "min": 4 }]*/
/*eslint-env es6*/

var { a: [x]} = {}; // Error: Identifier name `a` ...
```

And there is a PR working on to fix it (#12839).

 > Do we have an issue outlining our plan for this (just in case someone else needs to pick this work up)?

I will add an outlining plan for this on #12832. Thanks 👍

File: tests/lib/rules/id-length.js (hunk 7 above)
Comment (kaicataldo) [reply tests/lib/rules/id-length.js]: Thank you!

**Comment by kaicataldo** [comment]
Thanks for contributing!
